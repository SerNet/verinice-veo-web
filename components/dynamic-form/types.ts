/*
 * verinice.veo web
 * Copyright (C) 2022  Jonas Heitmann
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { JSONSchema7 } from 'json-schema';

export default {};

export interface IVeoFormsAdditionalContext {
  [pointer: string]: {
    objectSchema?: Record<string, any>;
    formSchema?: Record<string, any>;
  };
}

export interface IVeoFormsReactiveFormActions {
  [pointer: string]: ((
    newValue: any,
    oldValue: any,
    newObject: Record<string, any>,
    oldObject: Record<string, any>
  ) => Record<string, any>)[];
}

export interface IVeoFormsTranslations {
  [lang: string]: Record<string, any>;
}

export interface IVeoFormElementRule {
  visible: boolean;
  disabled: boolean;
}

export type IDynamicFormElementOptions = Record<string, any> &
  IVeoFormElementRule;

export interface IVeoFormElementDefaultProps {
  metaData: JSONSchema7;
  disabled: boolean;
  objectCreationDisabled: boolean;
  debug: boolean;
  [key: string]: any;
}

export interface IVeoFormsElementDefinition {
  code: string;
  name: { [lang: string]: string }; // Name of this component (used in FormSchemaEditor)
  description: { [lang: string]: string }; // Description describing this component (used in FormSchemaEditor)
  conditions?: (_props: IVeoFormElementDefaultProps) => boolean[]; // Every condition has to be truthy in order for this component to get displayed. Components with more truthy conditions will be displayed first.
  bias?: number; // Bias so this component gets choosen over other components with the same amount of truthy conditions
}

export interface IVeoFormElementFormSchemaRule {
  effect: 'HIDE' | 'SHOW' | 'DISABLE' | 'ENABLE';
  condition: {
    scope: string;
    schema: JSONSchema7;
  };
}

export interface IVeoFormElementFormSchema {
  type: string;
  options: IDynamicFormElementOptions;
  elements?: IVeoFormElementFormSchema[];
  rule?: IVeoFormElementFormSchemaRule;
}

export interface IVeoFormControlFormSchema extends IVeoFormElementFormSchema {
  scope: string;
  label: string;
}

export interface IVeoFormLabelFormSchema extends IVeoFormElementFormSchema {
  text: string;
}

export interface IVeoFormWidgetFormSchema extends IVeoFormElementFormSchema {
  name: string;
}
