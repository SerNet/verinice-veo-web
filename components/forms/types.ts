/*
 * verinice.veo web
 * Copyright (C) 2022  Jonas Heitmann
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { JSONSchema7 } from 'json-schema';
import { IBaseObject } from '~/lib/utils';

export default {};

export interface IVeoFormsAdditionalContext {
  [pointer: string]: {
    objectSchema?: IBaseObject;
    formSchema?: IBaseObject;
  };
}

export interface IVeoFormsReactiveFormActions {
  [pointer: string]: ((newValue: any, oldValue: any, newObject: IBaseObject, oldObject: IBaseObject) => IBaseObject)[];
}

export interface IVeoFormsTranslations {
  [lang: string]: IBaseObject;
}

export interface IVeoFormElementRule {
  visible: boolean;
  disabled: boolean;
}

export interface IVeoFormElementDefaultProps {
  metaData: JSONSchema7;
  disabled: boolean;
  objectCreationDisabled: boolean;
  translations: IBaseObject;
  debug: boolean;
  [key: string]: any;
}

export interface IVeoFormsElementDefinition {
  code: string;
  name: { [lang: string]: string }; // Name of this component (used in FormSchemaEditor)
  description: { [lang: string]: string }; // Description describing this component (used in FormSchemaEditor)
  conditions?: (_props: IVeoFormElementDefaultProps) => boolean[]; // Every condition has to be truthy in order for this component to get displayed. Components with more truthy conditions will be displayed first.
  bias?: number; // Bias so this component gets choosen over other components with the same amount of truthy conditions
}

export interface IVeoFormElementFormSchemaRule {
  effect: 'HIDE' | 'SHOW' | 'DISABLE' | 'ENABLE';
  condition: {
    scope: string;
    schema: JSONSchema7;
  };
}

export interface IVeoFormElementFormSchema {
  type: string;
  options: IBaseObject;
  elements?: IVeoFormElementFormSchema[];
  rule?: IVeoFormElementFormSchemaRule;
}

export interface IVeoFormControlFormSchema extends IVeoFormElementFormSchema {
  scope: string;
  label: string;
}

export interface IVeoFormLabelFormSchema extends IVeoFormElementFormSchema {
  text: string;
}

export interface IVeoFormWidgetFormSchema extends IVeoFormElementFormSchema {
  name: string;
}
