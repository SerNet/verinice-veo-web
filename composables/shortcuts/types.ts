/*
 * verinice.veo web
 * Copyright (C) 2025 Aziz Khalledi
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 */

export const SHORTCUT_DIALOG_DISPLAY = 'Shift + ?';
export const CATEGORY_DOMAIN_NAVIGATION = 'Domain Navigation';
export const SEQUENCE_TIMEOUT_MS = 1000;
export const MAX_SEQUENCE_LENGTH = 3;

export interface ShortcutAction {
  id: string;
  name: string;
  keys: string;
  description: string;
  action: () => void;
  category?: string;
  disabled?: boolean;
}

export interface ShortcutGroup {
  id: string;
  name: string;
  shortcuts: ShortcutAction[];
}

export interface TrieNode {
  shortcuts: ShortcutAction[];
  children: Map<string, TrieNode>;
}

export interface MatchResult {
  exactMatches: ShortcutAction[];
  hasPartialMatches: boolean;
}

export interface KeyboardHandlerOptions {
  onExactMatch: (shortcut: ShortcutAction) => void;
  onSequenceReset: () => void;
  isEnabled: () => boolean;
}

export interface ShortcutRegistry {
  register: (group: ShortcutGroup) => void;
  unregister: (id?: string) => void;
  clear: () => void;
}
