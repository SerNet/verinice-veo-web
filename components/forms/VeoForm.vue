<script lang="ts">
import Vue, { VNode, PropOptions } from 'vue'
import { JSONSchema7 } from 'json-schema'
import { JsonPointer } from 'json-ptr'

import vjp from 'vue-json-pointer'
import Ajv, { RequiredParams } from 'ajv'
import { chunk, merge } from 'lodash'
import { UISchema, UISchemaElement, UIRule } from '~/types/UISchema'
import { BaseObject, IApi } from '~/components/forms/utils'
import Label from '~/components/forms/Label.vue'
import Control from '~/components/forms/Control.vue'
import Layout from '~/components/forms/Layout.vue'
import Wrapper from '~/components/forms/Wrapper.vue'
import { IVeoFormSchemaTranslationCollectionItem, IVeoTranslation } from '~/types/VeoTypes'

interface IErrorMessageElement {
  pointer: string
  message: string
}

const ajv = new Ajv({
  allErrors: true,
  jsonPointers: true,
  strictKeywords: true
})

enum Mode {
  GENERAL = 'GENERAL',
  VEO = 'VEO'
}

interface IOptions {
  generator: {
    excludedProperties?: string[]
    elementsPerPage?: number
  }
}

export default Vue.extend({
  name: 'VeoForm',
  props: {
    value: {
      type: Object,
      default: () => ({})
    },
    schema: {
      type: Object,
      required: true
    } as PropOptions<JSONSchema7>,
    ui: {
      type: Object,
      default: undefined
    } as PropOptions<UISchema>,
    generalTranslation: {
      type: Object,
      default: () => {}
    } as PropOptions<IVeoTranslation>,
    customTranslation: {
      type: Object,
      default: () => {}
    } as PropOptions<IVeoFormSchemaTranslationCollectionItem>,
    options: {
      type: Object,
      default: undefined
    } as PropOptions<IOptions>,
    isValid: {
      type: Boolean
    },
    errorMessages: {
      type: Array,
      default: () => []
    } as PropOptions<IErrorMessageElement[]>,
    api: {
      type: Object,
      default: undefined
    } as PropOptions<IApi>
  },
  data() {
    return {
      page: 1,
      localSchema: this.schema,
      localUI: this.ui,
      defaultOptions: {
        generator: {
          excludedProperties: [
            '/id$',
            '/subType$',
            '/type$',
            '/domains$',
            '/owner$',
            '/href$',
            '/validFrom$',
            '/displayName$',
            '/resourcesUri$',
            '/searchesUri$',
            '/targetUri$',
            '/riskvalues$',
            '/assets$',
            '/extinguishingTime$',
            '^#/properties/links',
            '/applicableTo$',
            '/references$',
            '/updatedAt$',
            '/updatedBy$',
            '/createdAt$',
            '/createdBy$',
            '/parts$',
            '/members$',
            '/customAspects$',
            '/links$'
          ],
          elementsPerPage: -1
        }
      }
    }
  },
  computed: {
    pages(): UISchemaElement[] | undefined {
      if (this.localUI && this.localUI.elements) {
        return this.localUI.elements
          .filter(el => el.type === 'Layout' && el.options && el.options.format === 'page')
          .map((el, i) => ({
            ...el,
            options: { ...el.options, _pageID: i + 1 }
          }))
      }
      return undefined
    },
    visiblePages(): UISchemaElement[] | undefined {
      if (this.pages && this.pages.length >= 1) {
        return this.pages.filter(
          // TODO: it is a question, how Vue knows, what internally (rules) in function changes,
          // it works now, but it can probably cause BUGs, and then needs to pay ATTENTION!!!
          el => this.evaluateRule(el.rule).visible === true
        )
      }
      return undefined
    },
    pagesLength(): number {
      return this.visiblePages ? this.visiblePages.length : 1
    },
    currentPage(): UISchemaElement {
      if (this.localUI && this.localUI.elements && this.visiblePages) {
        return {
          ...this.localUI,
          elements: this.pagesLength > 1 ? [this.visiblePages[this.page - 1]] : [...this.localUI.elements]
        }
      } else {
        return { ...this.localUI }
      }
    },
    validate(): Ajv.ValidateFunction {
      return ajv.compile(this.localSchema)
    },
    valid(): boolean | PromiseLike<any> {
      return this.validate(this.value)
    },
    errorsMsgMap(): BaseObject {
      return !this.valid && this.validate.errors ? this.validate.errors.reduce(this.validationErrorTransform, {}) : {}
    },
    mergedOptions(): IOptions {
      return merge(this.defaultOptions, this.options)
    }
  },
  watch: {
    schema: {
      immediate: true,
      handler() {
        // IMPORTANT! This is needed to update localSchema when schema is updated
        // Else it cannot detect updated object of schema and does not update veo-form
        this.localSchema = JSON.parse(JSON.stringify(this.schema))
      }
    },
    ui: {
      immediate: true,
      // TODO: deep property is important to check for changes in FormSchema to update the form (import for live updates in FormSchemaEditor).
      // if it causes problems or performance issues, should be removed and another solution found
      deep: true,
      handler() {
        if (this.ui) {
          this.localUI = this.translate<UISchema>(this.ui)
        } else {
          this.localUI = this.translate<UISchema>(
            this.generateFormSchema(this.schema, this.mergedOptions.generator.excludedProperties, Mode.VEO)
          )
        }
      }
    },
    generalTranslation: {
      immediate: true,
      handler() {
        // this.localSchema = this.translate<JSONSchema7>(this.schema)
        if (this.ui) {
          this.localUI = this.translate<UISchema>(this.ui)
        } else if (this.localUI) {
          this.localUI = this.translate<UISchema>(this.localUI)
        }
      }
    },
    customTranslation: {
      immediate: true,
      handler() {
        // this.localSchema = this.translate<JSONSchema7>(this.schema)
        if (this.ui) {
          this.localUI = this.translate<UISchema>(this.ui)
        } else if (this.localUI) {
          this.localUI = this.translate<UISchema>(this.localUI)
        }
      }
    },
    visiblePages(newValue: UISchemaElement[] | undefined, oldValue: UISchemaElement[] | undefined) {
      // TODO: Refactor code to make it more readible and clear

      if (typeof oldValue !== 'undefined' && typeof newValue !== 'undefined') {
        // if both oldValue and newValue are defined, only then is important to k

        // get old current page to get old current page ID
        const oldCurrentPage = oldValue[this.page - 1]
        const oldCurrentPageID: number = oldCurrentPage.options && oldCurrentPage.options._pageID

        // Look in the new array of visible pages and find what the index of old current page is
        const currentPageIndexInNewValue = newValue.findIndex(
          el => el.options && el.options._pageID === oldCurrentPageID
        )
        // if the old current page was found in the new visible pages, adjust this.page, else this.page is 1
        this.page = currentPageIndexInNewValue !== -1 ? currentPageIndexInNewValue + 1 : 1
      } else {
        // else set the actual to 1 as default
        this.page = 1
      }
    },
    valid: {
      immediate: true,
      handler() {
        this.$emit('update:isValid', this.valid)
      }
    },
    errorsMsgMap: {
      immediate: true,
      handler() {
        this.$emit(
          'update:errorMessages',
          Object.entries(this.errorsMsgMap).map(([pointer, message]) => ({
            pointer,
            message
          }))
        )
      }
    }
  },
  created() {
    if (this.value && this.value.customAspects) {
      const customAspectsProperties = JsonPointer.get(this.localSchema, '#/properties/customAspects/properties') as any
      Object.keys(this.value.customAspects)
        .filter(property => typeof this.value.customAspects[property].type === 'undefined')
        .forEach(property => {
          const type = customAspectsProperties?.[property]?.properties?.type?.enum?.[0]
          if (type) {
            vjp.set(this.value, `/customAspects/${property}/type`, type)
          }
        })
    }
  },
  methods: {
    getLangText(langPointer: string): string {
      const translationKey = langPointer.replace('#lang/', '')
      return this.customTranslation?.[translationKey] || this.generalTranslation?.[translationKey] || ''
    },
    translate<T>(objectWithLangPointers: JSONSchema7 | UISchemaElement): T {
      return JSON.parse(
        JSON.stringify(objectWithLangPointers).replace(
          /"(#lang\/.*?)"/gi,
          (langMatchWithQuotes: string, langMatchWithoutQuotes: string) => {
            return JSON.stringify(this.getLangText(langMatchWithoutQuotes))
          }
        )
      )
    },
    propertyPath(path: string) {
      // TODO: Better translation from #/properties/name to #/name for values
      return String(path || '').replace(/\/properties\//g, '/')
    },
    evaluateRule(rule: UIRule | undefined) {
      const defaults = {
        visible: true,
        disabled: false
      }
      if (!rule) {
        return defaults
      }

      if (!['HIDE', 'SHOW', 'DISABLE', 'ENABLE'].includes(rule.effect)) {
        console.error(
          `Your rule effect "${rule.effect}" is not available!`,
          'Only these rule effects are permitted: "SHOW", "HIDE", "ENABLED", "DISABLED".'
        )
        return defaults
      }

      const v = JsonPointer.get(this.value, this.propertyPath(rule.condition.scope))

      // if rule condition is true
      if (ajv.validate(rule.condition.schema, v)) {
        switch (rule.effect) {
          case 'HIDE':
            return { ...defaults, visible: false }
          case 'SHOW':
            return { ...defaults, visible: true }
          case 'DISABLE':
            return { ...defaults, disabled: true }
          case 'ENABLE':
            return { ...defaults, disabled: false }
        }
      } else {
        // if rule condition is false
        // This means that SHOW and ENANBLE must be deactivated, because of above defined defaults
        switch (rule.effect) {
          case 'HIDE':
            return { ...defaults }
          case 'SHOW':
            return { ...defaults, visible: false }
          case 'DISABLE':
            return { ...defaults }
          case 'ENABLE':
            return { ...defaults, disabled: true }
        }
      }

      return defaults
    },
    setValue(scope: string, v: any) {
      if (scope) {
        // TODO: Here was changed JsonPointer with Vue.set() because of reactivity
        // Investigate how to work with it JsonPointer, because of JsonPaths
        // but have vue reactivity

        // console.log(this.value, scope, propertyPath(scope), v );
        // JsonPointer.set(this.value, propertyPath(scope), v, true);

        vjp.set(this.value, this.propertyPath(scope).replace('#/', '/'), v)
        this.$emit('input', this.value)
      }
    },
    validationErrorTransform(accummulator: {}, error: Ajv.ErrorObject) {
      const keyMatch = error.schemaPath.match(/((.+\/properties\/\w+\b)|(.+(?=\/required)))/g)
      if (!keyMatch) {
        throw new Error('Key does not match in Errors array')
      }

      const key =
        error.keyword !== 'required'
          ? keyMatch[0]
          : `${keyMatch[0]}/properties/${(error.params as RequiredParams).missingProperty}`

      return { ...accummulator, [key]: error.message }
    },
    generateControl(scope: string, items: BaseObject, mode: Mode = Mode.GENERAL): UISchemaElement {
      const propertyName = scope.split('/').pop()
      const label = propertyName ? (mode === Mode.VEO ? `#lang/${propertyName}` : propertyName) : ''
      return {
        type: 'Control',
        scope,
        options: {
          label
        },
        // Add property only if condition(here: items[scope]) is true https://stackoverflow.com/a/40560953/6072503
        ...(items[scope] && {
          elements: items[scope].map(this.generateControl)
        })
      }
    },
    generateGroups(content: UISchemaElement[], scopes: string[]) {
      const regCustomAspect = /#\/properties\/customAspects\/properties\/\w+/
      const uniqueCustomAspects = [
        ...new Set(
          scopes
            .filter(scope => scope.includes('#/properties/customAspects/properties'))
            .map(scope => {
              const matchedCustomAspect = scope.match(regCustomAspect)
              return matchedCustomAspect && matchedCustomAspect[0]
            })
        )
      ] as string[]

      return [
        ...content.filter((el: any) => el.scope && !regCustomAspect.test(el.scope)),
        ...uniqueCustomAspects.map(uniqueCustomAspect => {
          return {
            type: 'Layout',
            options: {
              type: 'group',
              direction: 'vertical',
              label: uniqueCustomAspect.split('/').pop()
            },
            elements: [...content.filter((el: any) => el.scope && el.scope.includes(uniqueCustomAspect))]
          }
        })
      ] as UISchemaElement[]
    },
    generatePages(content: UISchemaElement[]) {
      return chunk(content, this.mergedOptions.generator.elementsPerPage).map(
        (pageContent): UISchemaElement => ({
          type: 'Layout',
          options: {
            format: 'page'
          },
          elements: pageContent
        })
      )
    },
    generateFormSchema(objectSchema: JSONSchema7, excludedProperties: string[] = [], mode: Mode = Mode.GENERAL): any {
      const items: BaseObject = {}
      // @ts-ignore
      let schemaMap = Object.keys(JsonPointer.flatten(objectSchema, '#')) // TODO: Is '#' the right argument?
      const excludedPropertiesRegexp = excludedProperties.map(prop => new RegExp(prop))
      schemaMap =
        excludedPropertiesRegexp.length > 0
          ? schemaMap.filter(el => !excludedPropertiesRegexp.some(reg => reg.test(el)))
          : schemaMap
      const scopes = schemaMap
        .filter(el => /#\/(\w|\/)*properties\/\w+$/g.test(el))
        .filter((el, i, arr) => !arr.some(someEl => new RegExp(String.raw`${el}/properties/\w+`, 'g').test(someEl)))
        .filter(el => {
          if (/\/properties\/\w+\/items\/properties\/\w+$/g.test(el)) {
            const [parent, child] = el.split(/\/items(?=\/properties\/\w+$)/g)
            items[parent] = items[parent] ? [...items[parent], `#${child}`] : [`#${child}`]
            return false
          } else {
            return true
          }
        })
      let content = scopes.map(scope => this.generateControl(scope, items, mode))

      // Generate Groups for each customAspect
      content = this.generateGroups(content, scopes)

      content = this.mergedOptions.generator.elementsPerPage === -1 ? content : this.generatePages(content)
      return {
        type: 'Layout',
        options: {
          format: 'group',
          direction: 'vertical'
        },
        elements: content
      }
    }
  },
  render(h): VNode {
    const createComponent = (element: UISchemaElement, formSchemaPointer: string, elementLevel: number): VNode => {
      const createChildren = () => {
        return (
          element.elements &&
          element.elements.map((elem, index) =>
            createComponent(elem, `${formSchemaPointer}/elements/${index}`, elementLevel + 1)
          )
        )
      }

      const rule = this.evaluateRule(element.rule)

      switch (element.type) {
        case 'Layout':
          return h(
            Layout,
            {
              props: {
                options: element.options,
                formSchemaPointer,
                ...rule
              }
            },
            createChildren()
          )
        case 'Control': {
          let partOfProps: { [key: string]: any } = {
            name: undefined,
            schema: {},
            value: undefined,
            validation: {},
            generalTranslation: undefined,
            customTranslation: undefined,
            api: {}
          }

          if (element.scope) {
            const elementName = element.scope.split('/').pop() as string
            const elementSchema = JsonPointer.get(this.localSchema, element.scope) as any
            const elementValue = JsonPointer.get(this.value, this.propertyPath(element.scope)) as any
            const elementParentSchema = JsonPointer.get(this.localSchema, '#') as any
            const isRequired =
              Array.isArray(elementParentSchema.required) && elementParentSchema.required.includes(elementName)

            partOfProps = {
              name: elementName,
              schema: elementSchema,
              generalTranslation: this.generalTranslation,
              customTranslation: this.customTranslation,
              // TODO: Check InputNumber.vue or other Elements with "clear" and deafult value. Change how default value is used to fix bug
              value: typeof elementValue !== 'undefined' ? elementValue : elementSchema && elementSchema.default,
              validation: {
                objectSchema: {
                  errorMsg: this.errorsMsgMap[element.scope]
                }
              },
              api: this.api
            }
          }
          return h(Control, {
            props: {
              ...rule,
              elements: element.elements,
              options: element.options,
              ...partOfProps
            },
            on: {
              input: (v: any) => element.scope && this.setValue(element.scope, v),
              change: (v: any) => element.scope && this.setValue(element.scope, v)
            }
          })
        }
        case 'Label':
          return h(Label, {
            props: {
              ...rule,
              options: element.options,
              text: element.text
            }
          })
      }
    }

    return h(
      Wrapper,
      {
        props: {
          page: this.page,
          length: this.pagesLength
        },
        on: {
          input: (v: any) => (this.page = v)
        }
      },
      [createComponent(this.currentPage, '#', 0)]
    )
  }
})
</script>
