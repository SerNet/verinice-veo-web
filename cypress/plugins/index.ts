/// <reference types="cypress" />

const { initPlugin } = require('cypress-plugin-snapshots/plugin');

/**
 * @type {Cypress.PluginConfig}
 */
module.exports = (on, config) => {
  initPlugin(on, config);
  return config;
};
