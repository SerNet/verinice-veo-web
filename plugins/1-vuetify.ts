/*
 * verinice.veo web
 * Copyright (C) 2022  Jonas Heitmann
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { createVuetify } from 'vuetify';
import * as components from 'vuetify/components';
import * as directives from 'vuetify/directives';
import { aliases, mdi } from 'vuetify/iconsets/mdi-svg';
import { de, en } from 'vuetify/locale';
import 'vuetify/styles';
import Cookies from 'js-cookie';

// Vuetify Lab components
import { VDateInput } from 'vuetify/labs/VDateInput';

const vuetify = createVuetify({
  components: {
    ...components,
    VDateInput
  },
  directives,
  ssr: true,
  icons: {
    aliases,
    sets: {
      mdi
    },
    defaultSet: 'mdi' // This is already the default value - only for display purposes
  },
  locale: {
    // adapter: createVueI18nAdapter({ i18n, useI18n }) // Currently doesn't work with nuxt
    locale: Cookies.get('i18n_redirected') || navigator.language.split('-')[0],
    messages: { de, en }
  },
  theme: {
    defaultTheme: 'light',
    themes: {
      light: {
        colors: {
          primary: '#c90000',
          // secondary: '#',
          accent: '#ddd',
          surface: '#fff',
          color: '#000',
          basepage: '#f3f3f3'
          // error: '#',
          // info: '#'
          // success: '#',
          // warning: '#'
        },
        variables: {
          'disabled-opacity': 0.6
        }
      },
      dark: {
        colors: {
          primary: '#c90000',
          // secondary: '#',
          accent: '#000',
          surface: '#1c1c1c',
          color: '#fff',
          basepage: '#303030'
          // error: '#',
          // info: '#'
          // success: '#',
          // warning: '#'
        },
        variables: {
          'disabled-opacity': 0.6
        }
      }
    }
  }
});

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(vuetify);
});
