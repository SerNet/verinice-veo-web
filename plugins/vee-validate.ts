import Vue from 'vue'
import {
  ValidationProvider,
  extend
} from 'vee-validate'

export function install(vue: typeof Vue) {
  extend('objectSchema', {
    params: ['errorMsg'],
    validate: (_, args: Record<string, any>) => ({
      required: true,
      valid: !args.errorMsg
    }),
    message: (_, args) => args.errorMsg,
    computesRequired: true
  })
  vue.component('ValidationProvider', ValidationProvider)
}

export default function () {
  install(Vue)
}
