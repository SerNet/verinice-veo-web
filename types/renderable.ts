/*
 * verinice.veo web
 * Copyright (C) 2021  Markus Werner, Davit Svandize
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { UISchema } from '@/types/UISchema';
import { JSONSchema7 } from 'json-schema';

export interface Renderable {
  schema: JSONSchema7;
  ui: UISchema;
  value: {
    [propName: string]: any;
  };
}
