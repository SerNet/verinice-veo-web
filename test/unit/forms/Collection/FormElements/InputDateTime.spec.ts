import Vue from 'vue'
import { mount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import flushPromises from 'flush-promises'

import { install as VeeValidate } from '@/plugins/vee-validate'
import VeoForm from '~/components/forms/VeoForm.vue'
import { Renderable } from '~/types/renderable'
Vue.use(VeeValidate)
Vue.use(Vuetify)
const vuetify = new Vuetify()

describe('InputDateTime.vue', () => {
  it('should render input-date-time element for date-times', async() => {
    const form: Renderable = {
      schema: {
        properties: {
          dueDateTime: {
            type: 'string',
            format: 'date'
          }
        }
      },
      ui: {
        type: 'Layout',
        options: {
          direction: 'vertical',
          format: 'group'
        },
        elements: [
          {
            type: 'Control',
            scope: '#/properties/dueDateTime',
            options: {
              label: 'Due Date-Time'
            }
          }
        ]
      },
      value: {
        dueDateTime: '2020-08-05T12:05:00+02:00'
      }
    }

    const wrapper = mount(VeoForm, {
      vuetify,
      propsData: { ...form }
    })

    // Fixes immediate:true bugs with setProps() of vue test utils
    // https://github.com/vuejs/vue-test-utils/issues/1140#issuecomment-544156893
    wrapper.vm.$parent.$forceUpdate()
    await wrapper.vm.$nextTick()

    const inputElement = wrapper.find(
      '.vf-wrapper > .vf-layout > .vf-control .v-input.v-text-field input'
    ).element as HTMLInputElement

    expect(wrapper.element).toMatchSnapshot('Initial render')

    expect(inputElement.value).toBe('2020-08-05T12:05:00+02:00')

    wrapper.setProps({
      value: { dueDateTime: '1999-08-05T12:05:00+02:00' }
    })
    await flushPromises()

    expect(wrapper.element).toMatchSnapshot('Updated DateTime')

    expect(inputElement.value).toBe('1999-08-05T12:05:00+02:00')
  })
})
