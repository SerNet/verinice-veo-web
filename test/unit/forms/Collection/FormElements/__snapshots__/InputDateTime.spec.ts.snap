// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDateTime.vue should render input-date-time element for date-times: Initial render 1`] = `
<div
  class="vf-wrapper d-flex flex-column"
>
  <div
    class="vf-layout vf-group d-flex mx-auto flex-column"
    style="max-width: 800px; width: 100%;"
  >
    <div
      class="vf-control col"
    >
      <div>
        <span>
          <div
            class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-calendar theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label v-label--active theme--light"
                    for="input-10"
                    style="left: 0px; position: absolute;"
                  >
                    Due Date-Time
                  </label>
                  <input
                    id="input-10"
                    type="text"
                  />
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--clear"
                  >
                    <button
                      aria-label="clear icon"
                      class="v-icon notranslate v-icon--link mdi mdi-close theme--light"
                      type="button"
                    />
                  </div>
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <div
                    class="v-messages__wrapper"
                  />
                </div>
              </div>
            </div>
          </div>
        </span>
        <div
          class="v-menu"
        >
          <!---->
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`InputDateTime.vue should render input-date-time element for date-times: Updated DateTime 1`] = `
<div
  class="vf-wrapper d-flex flex-column"
>
  <div
    class="vf-layout vf-group d-flex mx-auto flex-column"
    style="max-width: 800px; width: 100%;"
  >
    <div
      class="vf-control col"
    >
      <div>
        <span>
          <div
            class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-calendar theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label v-label--active theme--light"
                    for="input-10"
                    style="left: 0px; position: absolute;"
                  >
                    Due Date-Time
                  </label>
                  <input
                    id="input-10"
                    type="text"
                  />
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--clear"
                  >
                    <button
                      aria-label="clear icon"
                      class="v-icon notranslate v-icon--link mdi mdi-close theme--light"
                      type="button"
                    />
                  </div>
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <div
                    class="v-messages__wrapper"
                  />
                </div>
              </div>
            </div>
          </div>
        </span>
        <div
          class="v-menu"
        >
          <!---->
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
